### PR TITLE
DEV: update `SequenceCollection` and `Alignment` construction functions

### PR DESCRIPTION
**Pull request by KatherineCaley**
## Summary by Sourcery

Refactor the SequenceCollection and Alignment constructor functions to support various input types and improve sequence name handling. Enhance the seq_to_gap_coords function to support new_sequence.Sequence objects and improve its return type. Add comprehensive tests to verify the correct functionality of these enhancements.

Enhancements:
- Refactor the SequenceCollection and Alignment constructor functions to handle various input types such as dict, list, and set, and improve the handling of sequence names and offsets.
- Improve the handling of sequence names by introducing a name_map to map between provided names and sequence names when constructing alignments.
- Enhance the seq_to_gap_coords function to support new_sequence.Sequence objects and improve its return type to include numpy arrays.

Tests:
- Add comprehensive tests for the SequenceCollection and Alignment constructor functions to ensure they handle different input types and scenarios, including dicts, lists, sets, and sequences with offsets.
- Introduce tests to verify the correct handling of sequence names and offsets in SequenceCollection and Alignment constructors.
- Add tests for the seq_to_gap_coords function to ensure it correctly processes sequences and returns expected gap coordinates.

### CODE REVIEW

File: src/cogent3/core/new_alignment.py [hunk 1]
```diff
@@ -4350,36 +4363,50 @@ def _mapped(self, slicemap):
...
 ) -> Alignment:
-    # todo: implement a coerce to AlignedSeqsData dict function
+    if len(data) == 0:
+        raise ValueError("data must be at least one sequence.")
+
     moltype = new_moltype.get_moltype(moltype)
     alphabet = moltype.most_degen_alphabet()
-    aligned_data = AlignedSeqsData.from_seqs(
+
+    if isinstance(data, dict) and isinstance(
```
[the remark this answers is not in the document]
Comment (GavinHuttley) [reply src/cogent3/core/new_alignment.py:4382]: @KatherineCaley I think this a good idea. This function can be used by both `make_(un)aligned_seqs()` functions as they're doing the same transformation. But the function name needs to be better.

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
such massive progress @KatherineCaley ! small change requested.

File: src/cogent3/core/new_alignment.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KatherineCaley) [reply src/cogent3/core/new_alignment.py]: I agree, done

**Review by GavinHuttley — APPROVED** [review]
Brillo!
For your next PR don't do the casting of sequences to `numpy.array` in `_make_name_map()`. Any conversions should happen in one place, the `coerce_to_seqs_data_dict()` function.

**Pull request merged**